src/arch/riscv/isa/formats/fp.isa

```diff
@@ -36,20 +36,20 @@ def template FloatExecute {{
     Fault %(class_name)s::execute(ExecContext *xc,
         trace::InstRecord *traceData) const
     {
-        STATUS status = xc->readMiscReg(MISCREG_STATUS);
-        if (status.fs == FPUStatus::OFF)
-            return std::make_shared<IllegalInstFault>("FPU is off", machInst);
-
-        status.fs = FPUStatus::DIRTY;
-        xc->setMiscReg(MISCREG_STATUS, status);
+        bool set_dirty = true;
+        Fault fault = updateFPUStatus(xc, machInst, set_dirty);
+        if (fault != NoFault) { return fault; }
 
         %(op_decl)s;
         %(op_rd)s;
 
+        RegVal FFLAGS = xc->readMiscReg(MISCREG_FFLAGS);
         std::feclearexcept(FE_ALL_EXCEPT);
         %(code)s;
-        xc->setMiscReg(MISCREG_FFLAGS_EXE, softfloat_exceptionFlags);
+
+        FFLAGS |= softfloat_exceptionFlags;
         softfloat_exceptionFlags = 0;
+        xc->setMiscReg(MISCREG_FFLAGS, FFLAGS & FFLAGS_MASK);
```

Comment:
If you do the change again. It will cause fflags issue in O3CPU https://github.com/gem5/gem5/issues/755

Reply:
Hey, thanks for the comment, we had doubts about that part. Could you briefly explain what the bug is here? I read #755  but from the traces given I do not yet understand the "logic" behind the bug. Should we set FFLAGS instead of OR'ing?

Reply:
It means we will get the wrong fflags result in O3CPU compared with AtomicCPU, you can test via command

```bash
build/RISCV/gem5.opt --debug-flags=Exec tests/gem5/configs/simple_binary_run.py rv64uf-ps-fadd o3 riscv --num-cores 4 --resource-directory tests/gem5/resources
```

For atomic CPU
```bash
build/RISCV/gem5.opt --debug-flags=Exec tests/gem5/configs/simple_binary_run.py rv64uf-ps-fadd atomic riscv --num-cores 4 --resource-directory tests/gem5/resources
```
> Should we set FFLAGS instead of OR'ing?

No, we accumulate the FFLAGS in each float arithmetic operation.

It's because the O3CPU[3] handles setMiscReg in different way compare to other CPU models, SimpleCPU[1] and MinorCPU[2]. The O3CPU wouldn't write the value back to MiscRegFile until the instruction is committed, we may get the old FFLAGS value in the next float instruction, and cause the riscv-tests failed. So that PRhttps://github.com/gem5/gem5/pull/868  use eatra misc_reg index `MISCREG_FFLAGS_EXE` to accumulate the FFLAGS to `MISCREG_FFLAGS` `miscRegFile`, just like what is implemented in the ARM[4]

[1] SimpleCPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/simple/exec_context.hh#L259-L263

[2] MinorCPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/minor/exec_context.hh#L263-L267

[3] O3CPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/o3/dyn_inst.hh#L1016-L1032

[4]